### PR TITLE
Rework user backend 

### LINKED ISSUE #81: Refactor the User backend to improve code structure and efficiency
Modify User constructor
Remove unnecessary attributes
Clean up repository functions
Enhance UserViewModel
Update affected tests

PS: "refactor" might be an understatement as this will comprise heavy changes

### PR DESCRIPTION
**Pull request by Armetiis**
This pull request implements significant refactoring of the User class, repository and viewModel to improve code structure, efficiency, and maintainability.

Closes #81 

##### User Constructor
Removed default arguments to ensure proper initialization

##### User Class Attributes
Deleted lastKnownLocation attribute
> Rationale: Storing outdated location data can lead to user confusion

Deleted homeAdress
> We can add this later if necessary, but not as the form of a string, as it is pretty much useless
##### Repository Cleanup
Removed the following functions:
- addFavouriteParking
- removeFavouriteParking
> These operations will be handled at the ViewModel level

##### UserViewModel Enhancements
Added selectedUser state
Implemented addFavoriteParkingToSelectedUser(parkingId: String) and removeFavoriteParkingFromSelectedUser(parkingId: String) for the selected user

### Test Suite Updates
Revised UserTest, UserRepositoryTest, and UserViewModelTest
Added new tests for ViewModel favorite parking operations

### Coverage
image

### Future Work
A subsequent PR will implement getFavouriteParkings and getContributedParkings functionality in the UserViewModel.

### CODE REVIEW

File: app/src/test/java/com/github/se/cyrcle/model/user/UserViewModelTest.kt [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+import org.mockito.MockitoAnnotations
+import org.mockito.kotlin.any
+import org.mockito.kotlin.eq
+import org.mockito.kotlin.verify
+import org.robolectric.RobolectricTestRunner
+
+@RunWith(RobolectricTestRunner::class)
+class UserViewModelTest {
+
+  @Mock private lateinit var userViewModel: UserViewModel
```
Comment (matttsss) [inline app/src/test/java/com/github/se/cyrcle/model/user/UserViewModelTest.kt:17]: The user viewModel that is tested should not be a mock. The annotation is not needed

File: app/src/test/java/com/github/se/cyrcle/model/user/UserViewModelTest.kt [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+  @Test
+  fun setCurrentUserTest() {
+    userViewModel.setCurrentUser(UserTestInstances.user1)
+
+    // Check if the user returned is the correct one
+    assert(userViewModel.currentUser.value == UserTestInstances.user1)
+  }
+
+  @Test
+  fun getUserByIdTest() = runBlocking {
```
Comment (matttsss) [inline app/src/test/java/com/github/se/cyrcle/model/user/UserViewModelTest.kt:42]: I am prety sure that the `runBlocking` is not usefull and can be removed. (To be confirmed by re-running the tests)

**Review by matttsss — COMMENTED** [review]
Thanks for cleanning up and documenting this package, it was well needed. If you could address my two points and it will be all good for me 👌

File: app/src/test/java/com/github/se/cyrcle/model/user/UserViewModelTest.kt (hunk 2 above)
Comment (Armetiis) [reply app/src/test/java/com/github/se/cyrcle/model/user/UserViewModelTest.kt:42]: You are right, thanks for the remark

File: app/src/test/java/com/github/se/cyrcle/model/user/UserViewModelTest.kt (hunk 1 above)
Comment (Armetiis) [reply app/src/test/java/com/github/se/cyrcle/model/user/UserViewModelTest.kt:17]: Indeed

File: app/src/main/java/com/github/se/cyrcle/model/user/UserViewModel.kt [hunk 3]
```diff
@@ -1,121 +1,114 @@
...
+    _currentUser.value = user
+  }
+
+  /**
+   * Gets a user by ID from the Firestore database and sets it as the current user.
+   *
+   * @param userId the ID of the user to get
+   * @return the user with the given ID
+   */
+  fun getUserById(userId: String) {
```
Comment (sean-prz) [inline app/src/main/java/com/github/se/cyrcle/model/user/UserViewModel.kt:30]: The documentation does not match the function

**Review by sean-prz — APPROVED** [review]
Nice rework ! 
I Only checked carefully the userViewModel and its unit test as the rest of the test will be refactored this sprint anyways.

**Review by matttsss — APPROVED** [review]

**Pull request merged**